File: exercises/go/user_repository_test.go
package main

import (
	"context"
	"database/sql"
	"testing"
	"time"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
	"github.com/testcontainers/testcontainers-go"
	"github.com/testcontainers/testcontainers-go/modules/postgres"
)

func TestUserRepository(t *testing.T) {
	ctx := context.Background()

	// Start PostgreSQL container
	postgresContainer, err := postgres.RunContainer(ctx,
		testcontainers.WithImage("postgres:15"),
		postgres.WithDatabase("testdb"),
		postgres.WithUsername("test"),
		postgres.WithPassword("test"),
		postgres.WithInitScripts("../../exercises/go/init.sql"),
	)
	require.NoError(t, err)
	defer postgresContainer.Terminate(ctx)

	// Get connection string
	connStr, err := postgresContainer.ConnectionString(ctx, "sslmode=disable")
	require.NoError(t, err)

	// Connect to database
	db, err := sql.Open("postgres", connStr)
	require.NoError(t, err)
	defer db.Close()

	// Create repository
	repo := NewUserRepository(db)

	// Test cleanup before each test
	cleanup := func() {
		repo.DeleteAllUsers()
	}

	t.Run("CreateUser", func(t *testing.T) {
		cleanup()

		user := NewUserWithNames("testuser", "test@example.com", "Test", "User")

		createdUser, err := repo.CreateUser(user)
		require.NoError(t, err)
		assert.NotNil(t, createdUser)
		assert.NotZero(t, createdUser.ID)
		assert.Equal(t, "testuser", createdUser.Username)
		assert.Equal(t, "test@example.com", createdUser.Email)
		assert.Equal(t, "Test", createdUser.FirstName)
		assert.Equal(t, "User", createdUser.LastName)
		assert.False(t, createdUser.CreatedAt.IsZero())
		assert.False(t, createdUser.UpdatedAt.IsZero())
	})

	t.Run("FindByID", func(t *testing.T) {
		cleanup()

		user := NewUserWithNames("finduser", "find@example.com", "Find", "User")
		createdUser, err := repo.CreateUser(user)
		require.NoError(t, err)

		foundUser, err := repo.FindByID(createdUser.ID)
		require.NoError(t, err)
		assert.NotNil(t, foundUser)
		assert.Equal(t, "finduser", foundUser.Username)
		assert.Equal(t, "find@example.com", foundUser.Email)
	})

	t.Run("FindByUsername", func(t *testing.T) {
		cleanup()

		user := NewUserWithNames("usernameuser", "username@example.com", "Username", "User")
		_, err := repo.CreateUser(user)
		require.NoError(t, err)

		foundUser, err := repo.FindByUsername("usernameuser")
		require.NoError(t, err)
		assert.NotNil(t, foundUser)
		assert.Equal(t, "username@example.com", foundUser.Email)
	})

	t.Run("FindByEmail", func(t *testing.T) {
		cleanup()

		user := NewUserWithNames("emailuser", "email@example.com", "Email", "User")
		_, err := repo.CreateUser(user)
		require.NoError(t, err)

		foundUser, err := repo.FindByEmail("email@example.com")
		require.NoError(t, err)
		assert.NotNil(t, foundUser)
		assert.Equal(t, "emailuser", foundUser.Username)
	})

	t.Run("UpdateUser", func(t *testing.T) {
		cleanup()

		user := NewUserWithNames("updateuser", "update@example.com", "Update", "User")
		createdUser, err := repo.CreateUser(user)
		require.NoError(t, err)

		originalUpdatedAt := createdUser.UpdatedAt

		// Wait a bit to ensure timestamp difference
		time.Sleep(10 * time.Millisecond)

		createdUser.FirstName = "Updated"
		createdUser.LastName = "Name"

		err = repo.UpdateUser(createdUser)
		require.NoError(t, err)

		assert.Equal(t, "Updated", createdUser.FirstName)
		assert.Equal(t, "Name", createdUser.LastName)
		assert.True(t, createdUser.UpdatedAt.After(originalUpdatedAt))
	})

	t.Run("DeleteUser", func(t *testing.T) {
		cleanup()

		user := NewUserWithNames("deleteuser", "delete@example.com", "Delete", "User")
		createdUser, err := repo.CreateUser(user)
		require.NoError(t, err)

		err = repo.DeleteUser(createdUser.ID)
		require.NoError(t, err)

		foundUser, err := repo.FindByID(createdUser.ID)
		require.NoError(t, err)
		assert.Nil(t, foundUser)
	})

	t.Run("FindAllUsers", func(t *testing.T) {
		cleanup()

		user1 := NewUserWithNames("user1", "user1@example.com", "User", "One")
		user2 := NewUserWithNames("user2", "user2@example.com", "User", "Two")

		_, err := repo.CreateUser(user1)
		require.NoError(t, err)
		_, err = repo.CreateUser(user2)
		require.NoError(t, err)

		users, err := repo.FindAllUsers()
		require.NoError(t, err)
		assert.Len(t, users, 2)

		usernames := make([]string, len(users))
		for i, user := range users {
			usernames[i] = user.Username
		}
		assert.Contains(t, usernames, "user1")
		assert.Contains(t, usernames, "user2")
	})

	t.Run("CountUsers", func(t *testing.T) {
		cleanup()

		count, err := repo.CountUsers()
		require.NoError(t, err)
		assert.Equal(t, 0, count)

		user1 := NewUserWithNames("countuser1", "count1@example.com", "Count", "User1")
		_, err = repo.CreateUser(user1)
		require.NoError(t, err)

		count, err = repo.CountUsers()
		require.NoError(t, err)
		assert.Equal(t, 1, count)

		user2 := NewUserWithNames("countuser2", "count2@example.com", "Count", "User2")
		_, err = repo.CreateUser(user2)
		require.NoError(t, err)

		count, err = repo.CountUsers()
		require.NoError(t, err)
		assert.Equal(t, 2, count)
	})

	t.Run("HandleNonExistentUser", func(t *testing.T) {
		cleanup()

		foundUser, err := repo.FindByID(999)
		require.NoError(t, err)
		assert.Nil(t, foundUser)
	})

	t.Run("PreventDuplicateUsernames", func(t *testing.T) {
		cleanup()

		user1 := NewUserWithNames("duplicate", "user1@example.com", "User", "One")
		user2 := NewUserWithNames("duplicate", "user2@example.com", "User", "Two")

		_, err := repo.CreateUser(user1)
		require.NoError(t, err)

		_, err = repo.CreateUser(user2)
		assert.Error(t, err)
		assert.Contains(t, err.Error(), "duplicate key value violates unique constraint")
	})

	t.Run("PreventDuplicateEmails", func(t *testing.T) {
		cleanup()

		user1 := NewUserWithNames("user1", "duplicate@example.com", "User", "One")
		user2 := NewUserWithNames("user2", "duplicate@example.com", "User", "Two")

		_, err := repo.CreateUser(user1)
		require.NoError(t, err)

		_, err = repo.CreateUser(user2)
		assert.Error(t, err)
		assert.Contains(t, err.Error(), "duplicate key value violates unique constraint")
	})

	t.Run("CompleteCRUDWorkflow", func(t *testing.T) {
		cleanup()

		// Create
		user := NewUserWithNames("cruduser", "crud@example.com", "CRUD", "User")
		createdUser, err := repo.CreateUser(user)
		require.NoError(t, err)
		assert.NotZero(t, createdUser.ID)

		// Read
		foundUser, err := repo.FindByID(createdUser.ID)
		require.NoError(t, err)
		assert.NotNil(t, foundUser)
		assert.Equal(t, "cruduser", foundUser.Username)

		// Update
		foundUser.FirstName = "Updated"
		err = repo.UpdateUser(foundUser)
		require.NoError(t, err)
		assert.Equal(t, "Updated", foundUser.FirstName)

		// Delete
		err = repo.DeleteUser(foundUser.ID)
		require.NoError(t, err)

		// Verify deletion
		deletedUser, err := repo.FindByID(foundUser.ID)
		require.NoError(t, err)
		assert.Nil(t, deletedUser)
	})
}

File: exercises/nodejs/src/__tests__/user-repository.test.ts
import { Pool } from 'pg';
import { StartedPostgreSqlContainer, PostgreSqlContainer } from '@testcontainers/postgres';
import { UserEntity, UserRepository } from '../user-repository';

describe('UserRepository', () => {
  let postgresContainer: StartedPostgreSqlContainer;
  let pool: Pool;
  let repository: UserRepository;

  beforeAll(async () => {
    // Start PostgreSQL container
    postgresContainer = await new PostgreSqlContainer('postgres:15')
      .withDatabase('testdb')
      .withUsername('test')
      .withPassword('test')
      .withInitScript('../../exercises/nodejs/init.sql')
      .start();

    // Create connection pool
    pool = new Pool({
      host: postgresContainer.getHost(),
      port: postgresContainer.getPort(),
      database: postgresContainer.getDatabase(),
      user: postgresContainer.getUsername(),
      password: postgresContainer.getPassword(),
    });

    repository = new UserRepository(pool);
  });

  afterAll(async () => {
    if (pool) {
      await pool.end();
    }
    if (postgresContainer) {
      await postgresContainer.stop();
    }
  });

  beforeEach(async () => {
    // Clean up test data before each test
    await repository.deleteAllUsers();
  });

  describe('createUser', () => {
    it('should create a new user successfully', async () => {
      const user = UserEntity.createWithNames(
        'testuser',
        'test@example.com',
        'Test',
        'User'
      );

      const createdUser = await repository.createUser(user);

      expect(createdUser).toBeDefined();
      expect(createdUser.id).toBeDefined();
      expect(createdUser.username).toBe('testuser');
      expect(createdUser.email).toBe('test@example.com');
      expect(createdUser.firstName).toBe('Test');
      expect(createdUser.lastName).toBe('User');
      expect(createdUser.createdAt).toBeDefined();
      expect(createdUser.updatedAt).toBeDefined();
    });

    it('should prevent duplicate usernames', async () => {
      const user1 = UserEntity.createWithNames(
        'duplicate',
        'user1@example.com',
        'User',
        'One'
      );
      const user2 = UserEntity.createWithNames(
        'duplicate',
        'user2@example.com',
        'User',
        'Two'
      );

      await repository.createUser(user1);

      await expect(repository.createUser(user2)).rejects.toThrow();
    });

    it('should prevent duplicate emails', async () => {
      const user1 = UserEntity.createWithNames(
        'user1',
        'duplicate@example.com',
        'User',
        'One'
      );
      const user2 = UserEntity.createWithNames(
        'user2',
        'duplicate@example.com',
        'User',
        'Two'
      );

      await repository.createUser(user1);

      await expect(repository.createUser(user2)).rejects.toThrow();
    });
  });

  describe('findById', () => {
    it('should find user by ID', async () => {
      const user = UserEntity.createWithNames(
        'finduser',
        'find@example.com',
        'Find',
        'User'
      );
      const createdUser = await repository.createUser(user);

      const foundUser = await repository.findById(createdUser.id!);

      expect(foundUser).toBeDefined();
      expect(foundUser!.username).toBe('finduser');
      expect(foundUser!.email).toBe('find@example.com');
    });

    it('should return null for non-existent user', async () => {
      const foundUser = await repository.findById(999);
      expect(foundUser).toBeNull();
    });
  });

  describe('findByUsername', () => {
    it('should find user by username', async () => {
      const user = UserEntity.createWithNames(
        'usernameuser',
        'username@example.com',
        'Username',
        'User'
      );
      await repository.createUser(user);

      const foundUser = await repository.findByUsername('usernameuser');

      expect(foundUser).toBeDefined();
      expect(foundUser!.email).toBe('username@example.com');
    });

    it('should return null for non-existent username', async () => {
      const foundUser = await repository.findByUsername('nonexistent');
      expect(foundUser).toBeNull();
    });
  });

  describe('findByEmail', () => {
    it('should find user by email', async () => {
      const user = UserEntity.createWithNames(
        'emailuser',
        'email@example.com',
        'Email',
        'User'
      );
      await repository.createUser(user);

      const foundUser = await repository.findByEmail('email@example.com');

      expect(foundUser).toBeDefined();
      expect(foundUser!.username).toBe('emailuser');
    });

    it('should return null for non-existent email', async () => {
      const foundUser = await repository.findByEmail('nonexistent@example.com');
      expect(foundUser).toBeNull();
    });
  });

  describe('updateUser', () => {
    it('should update user successfully', async () => {
      const user = UserEntity.createWithNames(
        'updateuser',
        'update@example.com',
        'Update',
        'User'
      );
      const createdUser = await repository.createUser(user);
      const originalUpdatedAt = createdUser.updatedAt;

      // Wait a bit to ensure timestamp difference
      await new Promise(resolve => setTimeout(resolve, 10));

      createdUser.firstName = 'Updated';
      createdUser.lastName = 'Name';

      const updatedUser = await repository.updateUser(createdUser);

      expect(updatedUser.firstName).toBe('Updated');
      expect(updatedUser.lastName).toBe('Name');
      expect(updatedUser.updatedAt!.getTime()).toBeGreaterThan(originalUpdatedAt!.getTime());
    });

    it('should throw error for non-existent user', async () => {
      const user = new UserEntity();
      user.id = 999;
      user.username = 'nonexistent';
      user.email = 'nonexistent@example.com';

      await expect(repository.updateUser(user)).rejects.toThrow();
    });
  });

  describe('deleteUser', () => {
    it('should delete user successfully', async () => {
      const user = UserEntity.createWithNames(
        'deleteuser',
        'delete@example.com',
        'Delete',
        'User'
      );
      const createdUser = await repository.createUser(user);

      const deleted = await repository.deleteUser(createdUser.id!);

      expect(deleted).toBe(true);

      const foundUser = await repository.findById(createdUser.id!);
      expect(foundUser).toBeNull();
    });

    it('should return false for non-existent user', async () => {
      const deleted = await repository.deleteUser(999);
      expect(deleted).toBe(false);
    });
  });

  describe('findAllUsers', () => {
    it('should return all users', async () => {
      const user1 = UserEntity.createWithNames(
        'user1',
        'user1@example.com',
        'User',
        'One'
      );
      const user2 = UserEntity.createWithNames(
        'user2',
        'user2@example.com',
        'User',
        'Two'
      );

      await repository.createUser(user1);
      await repository.createUser(user2);

      const users = await repository.findAllUsers();

      expect(users).toHaveLength(2);
      expect(users.map(u => u.username)).toEqual(
        expect.arrayContaining(['user1', 'user2'])
      );
    });

    it('should return empty array when no users exist', async () => {
      const users = await repository.findAllUsers();
      expect(users).toHaveLength(0);
    });
  });

  describe('countUsers', () => {
    it('should count users correctly', async () => {
      expect(await repository.countUsers()).toBe(0);

      const user1 = UserEntity.createWithNames(
        'countuser1',
        'count1@example.com',
        'Count',
        'User1'
      );
      await repository.createUser(user1);
      expect(await repository.countUsers()).toBe(1);

      const user2 = UserEntity.createWithNames(
        'countuser2',
        'count2@example.com',
        'Count',
        'User2'
      );
      await repository.createUser(user2);
      expect(await repository.countUsers()).toBe(2);
    });
  });

  describe('executeTransaction', () => {
    it('should execute transaction successfully', async () => {
      const result = await repository.executeTransaction(async (client) => {
        const user1 = UserEntity.createWithNames(
          'transuser1',
          'trans1@example.com',
          'Trans',
          'User1'
        );
        const user2 = UserEntity.createWithNames(
          'transuser2',
          'trans2@example.com',
          'Trans',
          'User2'
        );

        // Create users within transaction
        const createdUser1 = await repository.createUser(user1);
        const createdUser2 = await repository.createUser(user2);

        return [createdUser1, createdUser2];
      });

      expect(result).toHaveLength(2);
      expect(await repository.countUsers()).toBe(2);
    });

    it('should rollback transaction on error', async () => {
      await expect(
        repository.executeTransaction(async (client) => {
          const user1 = UserEntity.createWithNames(
            'rollbackuser1',
            'rollback1@example.com',
            'Rollback',
            'User1'
          );
          await repository.createUser(user1);

          // This should cause an error and rollback
          throw new Error('Transaction error');
        })
      ).rejects.toThrow('Transaction error');

      expect(await repository.countUsers()).toBe(0);
    });
  });

  describe('Complete CRUD workflow', () => {
    it('should handle complete CRUD operations', async () => {
      // Create
      const user = UserEntity.createWithNames(
        'cruduser',
        'crud@example.com',
        'CRUD',
        'User'
      );
      const createdUser = await repository.createUser(user);
      expect(createdUser.id).toBeDefined();

      // Read
      const foundUser = await repository.findById(createdUser.id!);
      expect(foundUser).toBeDefined();
      expect(foundUser!.username).toBe('cruduser');

      // Update
      foundUser!.firstName = 'Updated';
      const updatedUser = await repository.updateUser(foundUser!);
      expect(updatedUser.firstName).toBe('Updated');

      // Delete
      const deleted = await repository.deleteUser(updatedUser.id!);
      expect(deleted).toBe(true);

      // Verify deletion
      const deletedUser = await repository.findById(updatedUser.id!);
      expect(deletedUser).toBeNull();
    });
  });
});
